src/sentry/utils/outcomes.py

```diff
@@ -102,6 +104,13 @@ def track_outcome(
         billing_config["real_topic_name"] if use_billing else outcomes_config["real_topic_name"]
     )
 
+    # Generate idempotency key using message fields
+    # This is to ensure that the message is processed at exactly once; even if the outcome is reprocessed again.
+    # For billing purposes, this lets us avoid duplicate usage counts.
+    random_str = uuid4().hex
+    idempotency_key_fields = f"{random_str}|{event_id}|{org_id}|{project_id}|{key_id}|{outcome.value}|{category}|{reason}|{quantity}|{timestamp}"
+    idempotency_key = hashlib.sha256(idempotency_key_fields.encode("utf-8")).hexdigest()
```

Comment:
Couldn't we use the message partition/offset as a natural idempotency key? That would also let us have only a single key per topic/partition that acts as a watermark instead of having to set a key per outcome.

Reply:
@markstory I've implemented this alternative idea in https://github.com/getsentry/getsentry/pull/15332.